x/evm/keeper/state_transition.go

```diff
@@ -446,12 +481,78 @@ func (k *Keeper) ApplyMessageWithConfig(
 	// reset leftoverGas, to be used by the tracer
 	leftoverGas = msg.GasLimit - gasUsed
 
+	var replyLog StateDBGetLogsReply
+	sgxRpcClient.StateDBGetLogs(StateDBGetLogsArgs{}, &replyLog)
+
 	return &types.MsgEthereumTxResponse{
 		GasUsed:   gasUsed,
 		VmError:   vmError,
 		Ret:       ret,
-		Logs:      types.NewLogsFromEth(stateDB.Logs()),
+		Logs:      types.NewLogsFromEth(replyLog.Logs),
 		Hash:      cfg.TxConfig.TxHash.Hex(),
 		BlockHash: ctx.HeaderHash(),
 	}, nil
 }
+
+// prepareTxForSgx prepares the transaction for the SGX enclave. It:
+//   - creates an RPC server around the keeper to receive requests sent by the
+//     SGX
+//   - sends a "PrepareTx" request to the SGX enclave with the relevant tx and
+//     block info
+func (k *Keeper) prepareTxForSgx(ctx sdk.Context, msg core.Message, cfg *EVMConfig, sgxRpcClient *sgxRpcClient) error {
+	// Step 1. Create an RPC server to receive requests from the SGX enclave.
+	err := k.runRPCServer(ctx, msg, cfg)
+	if err != nil {
+		return err
+	}
+
+	// Step 2. Send a "PrepareTx" request to the SGX enclave.
+	chainConfigJson, err := json.Marshal(cfg.ChainConfig)
+	if err != nil {
+		return err
+	}
+	ctx.HeaderHash()
+	args := PrepareTxArgs{
+		Header: ctx.BlockHeader(),
+		Msg:    msg,
+		EvmConfig: PrepareTxEVMConfig{
+			ChainConfigJson: chainConfigJson,
+			CoinBase:        cfg.CoinBase,
+			BaseFee:         cfg.BaseFee,
+			TxConfig:        cfg.TxConfig,
+			DebugTrace:      cfg.DebugTrace,
+			NoBaseFee:       cfg.FeeMarketParams.NoBaseFee,
+			EvmDenom:        cfg.Params.EvmDenom,
+			Overrides:       ToGOB64(cfg.Overrides),
```

Comment:
Can you explain why we need a custom encoding here?

Reply:
@amaury1093 cfg.Overried isn't serializable.

Reply:
I'm not 100% ToGob64 works. This test fails. Could you double check on your side too @kenta-mori3322?

```go
func TestABC(t *testing.T) {
	a := types.StateOverride{}
	myNum := hexutil.Big(*big.NewInt(20))
	pMyNum := &myNum
	addr := common.HexToAddress("0x9F907cD9429939D2b41CED2950C996b9D8481B32")
	a[addr] = types.OverrideAccount{
		Balance: &pMyNum,
	}
	str := keeper.ToGOB64(&a)
	fmt.Println(str)
	b := keeper.FromGOB64(str)
	require.Equal(t, a, b)
}
```

I had the same issue with `cfg.ChainConfig`, and taking a look at geth code they support well JSON serialization. It seems that this OverrideAccount is also json-serialiazable, maybe we can switch to that?

Reply:
@amaury1093 Gotcha, let me take a look at it.